File: src/main/java/by/home/acs/language/archive/BaseArchiveFileSystem.java
package by.home.acs.language.archive;

import com.intellij.openapi.util.io.FileUtil;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.openapi.vfs.LocalFileProvider;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.vfs.newvfs.ArchiveFileSystem;
import com.intellij.openapi.vfs.newvfs.VfsImplUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public abstract class BaseArchiveFileSystem extends ArchiveFileSystem {
    protected String fileSystemProtocol;
    protected String fileSystemSeparator;

    //used instead of deprecated getLocalVirtualFileFor method
    @Override
    public @Nullable VirtualFile getLocalByEntry(@NotNull VirtualFile entry) {
        return super.getLocalByEntry(entry);
    }

    //used instead of deprecated findLocalVirtualFileByPath
    @Override
    public @Nullable VirtualFile findFileByPath(@NotNull String path) {
        return VfsImplUtil.findFileByPath(this, path);
    }

    @Override
    protected @Nullable String normalize(@NotNull String path) {
        var separatorIndex = path.indexOf(fileSystemSeparator);
        if (separatorIndex > 0) {
            var root = path.substring(0, separatorIndex);
            return FileUtil.normalize(root) + path.substring(separatorIndex);
        }
        return super.normalize(path);
    }

    @Override
    public @NotNull String getProtocol() {
        return fileSystemProtocol;
    }

    @Override
    protected @NotNull String extractRootPath(@NotNull String normalizedPath) {
        var separatorIndex = normalizedPath.indexOf(fileSystemSeparator);
        assert separatorIndex > 0;
        return normalizedPath.substring(0, separatorIndex + fileSystemSeparator.length());
    }

    @Override
    protected @NotNull String extractLocalPath(@NotNull String rootPath) {
        return StringUtil.trimEnd(rootPath, fileSystemSeparator);
    }

    @Override
    protected @NotNull String composeRootPath(@NotNull String localPath) {
        return localPath + fileSystemSeparator;
    }

    @Override
    public @Nullable VirtualFile findFileByPathIfCached(@NotNull String path) {
        return VfsImplUtil.findFileByPathIfCached(this, path);
    }

    @Override
    public @Nullable VirtualFile refreshAndFindFileByPath(@NotNull String path) {
        return VfsImplUtil.refreshAndFindFileByPath(this, path);
    }

    @Override
    public void refresh(boolean asynchronous) {
        VfsImplUtil.refresh(this, asynchronous);
    }

    @Override
    public @NotNull String extractPresentableUrl(@NotNull String path) {
        return StringUtil.trimEnd(path, fileSystemSeparator);
    }


}

File: src/main/kotlin/by/home/acs/language/archive/BaseArchiveFileSystem.kt
package by.home.acs.language.archive

import by.home.acs.language.archive.util.FileStructureUtil
import com.intellij.openapi.util.io.FileUtil
import com.intellij.openapi.util.text.StringUtil
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.vfs.newvfs.ArchiveFileSystem
import com.intellij.openapi.vfs.newvfs.VfsImplUtil

abstract class BaseArchiveFileSystem(
    private val fsProtocol: String,
    private val fsSeparator: String = FileStructureUtil.FS_SEPARATOR
) : ArchiveFileSystem() {
    private fun getVirtualFileForArchive(entryFile: VirtualFile?): VirtualFile? {
        return if (entryFile == null) null else getLocalByEntry(entryFile)
    }

    fun getArchiveRootForLocalFile(file: VirtualFile): VirtualFile? {
        return getRootByLocal(file)
    }

    override fun getProtocol() = fsProtocol

//    override fun isCorrectFileType(local: VirtualFile): Boolean {
//        return FileTypeRegistry.getInstance().getFileTypeByFileName(local.name) is fileType
//    }

    override fun extractPresentableUrl(path: String) = super.extractPresentableUrl(
        StringUtil.trimEnd(path, fsSeparator)
    )

    override fun normalize(path: String): String? {
        val separatorIndex = path.indexOf(fsSeparator)
        if (separatorIndex > 0) {
            val root = path.substring(0, separatorIndex)
            return FileUtil.normalize(root) + path.substring(separatorIndex)
        }
        return super.normalize(path)
    }

    override fun extractRootPath(path: String): String {
        val separatorIndex = path.indexOf(fsSeparator)
        assert(
            separatorIndex >= 0
        ) { "Path passed to ${this.javaClass.simpleName} must have separator '!/': $path" }
        return path.substring(0, separatorIndex + fsSeparator.length)
    }

    override fun extractLocalPath(rootPath: String): String {
        return StringUtil.trimEnd(rootPath, fsSeparator)
    }

    override fun composeRootPath(localPath: String): String {
        return localPath + fsSeparator
    }

    override fun findFileByPath(path: String): VirtualFile? {
        return VfsImplUtil.findFileByPath(this, path)
    }

    override fun findFileByPathIfCached(path: String): VirtualFile? {
        return VfsImplUtil.findFileByPathIfCached(this, path)
    }

    override fun refreshAndFindFileByPath(path: String): VirtualFile? {
        return VfsImplUtil.refreshAndFindFileByPath(this, path)
    }

    override fun refresh(asynchronous: Boolean) {
        VfsImplUtil.refresh(this, asynchronous)
    }
}
